File: push/src/main/java/com/nihility/XMPushUtils.java
package com.nihility;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.xiaomi.channel.commonutils.reflect.JavaCalls;
import com.xiaomi.mipush.sdk.PushContainerHelper;
import com.xiaomi.push.service.MIPushEventProcessor;
import com.xiaomi.xmpush.thrift.ActionType;
import com.xiaomi.xmpush.thrift.PushMetaInfo;
import com.xiaomi.xmpush.thrift.XmPushActionContainer;
import com.xiaomi.xmpush.thrift.XmPushActionNotification;
import com.xiaomi.xmpush.thrift.XmPushThriftSerializeUtils;

import org.apache.thrift.TBase;

import top.trumeet.common.utils.CustomConfiguration;
import top.trumeet.common.utils.Utils;

public class XMPushUtils {
    public static @NonNull CustomConfiguration getConfiguration(XmPushActionContainer container) {
        if (container == null) {
            return new CustomConfiguration(null);
        }
        return getConfiguration(container.getMetaInfo());
    }

    public static @NonNull CustomConfiguration getConfiguration(PushMetaInfo metaInfo) {
        if (metaInfo == null) {
            return new CustomConfiguration(null);
        }
        return new CustomConfiguration(metaInfo.getExtra());
    }

    public static @Nullable XmPushActionContainer packToContainer(byte[] payload) {
        return MIPushEventProcessor.buildContainer(payload);
    }

    public static @NonNull XmPushActionContainer packToContainer(XmPushActionNotification action, String packageName) {
        return packToContainer(action, packageName, ActionType.Notification, action.appId);
    }

    public static @NonNull <T extends TBase<T, ?>> XmPushActionContainer packToContainer(T action, String packageName, ActionType actionType, String appId) {
        return JavaCalls.callStaticMethod(
                PushContainerHelper.class.getName(), "generateRequestContainer",
                Utils.getApplication(), action, actionType,
                false, packageName, appId);
    }

    public static <T extends TBase<T, ?>> byte[] packToBytes(T container) {
        return XmPushThriftSerializeUtils.convertThriftObjectToBytes(container);
    }
}

File: push/src/main/java/com/nihility/XMPushUtils.kt
@file:Suppress("DEPRECATION")
package com.nihility
// Compatibility shim: legacy namespace forwarding to com.magisk317.*

import com.xiaomi.xmpush.thrift.ActionType
import com.xiaomi.xmpush.thrift.PushMetaInfo
import com.xiaomi.xmpush.thrift.XmPushActionContainer
import com.xiaomi.xmpush.thrift.XmPushActionNotification
import org.apache.thrift.TBase
import top.trumeet.common.utils.CustomConfiguration

object XMPushUtils {
    @JvmStatic
    fun getConfiguration(container: XmPushActionContainer?): CustomConfiguration =
        com.magisk317.XMPushUtils.getConfiguration(container)

    @JvmStatic
    fun getConfiguration(metaInfo: PushMetaInfo?): CustomConfiguration =
        com.magisk317.XMPushUtils.getConfiguration(metaInfo)

    @JvmStatic
    fun packToContainer(payload: ByteArray?): XmPushActionContainer? =
        com.magisk317.XMPushUtils.packToContainer(payload)

    @JvmStatic
    fun packToContainer(
        action: XmPushActionNotification,
        packageName: String
    ): XmPushActionContainer =
        com.magisk317.XMPushUtils.packToContainer(action, packageName)

    @JvmStatic
    fun <T : TBase<T, *>> packToContainer(
        action: T,
        packageName: String,
        actionType: ActionType,
        appId: String?
    ): XmPushActionContainer =
        com.magisk317.XMPushUtils.packToContainer(action, packageName, actionType, appId)

    @JvmStatic
    fun <T : TBase<T, *>> packToBytes(container: T): ByteArray =
        com.magisk317.XMPushUtils.packToBytes(container)
}
